File: solutions/python/custom-set/5/custom_set.py
class CustomSet:
    def __init__(self, elements=None):
        if elements is None:
            elements = []
        self.elements = set(elements)

    def isempty(self):
        return len(self.elements) == 0

    def __contains__(self, element):
        return element in self.elements

    def issubset(self, other):
        return self.elements.issubset(other.elements)

    def isdisjoint(self, other):
        return self.elements.isdisjoint(other.elements)

    def __eq__(self, other):
        if not isinstance(other, CustomSet):
            return False
        return self.elements == other.elements

    def add(self, element):
        self.elements.add(element)

    def intersection(self, other):
        return CustomSet(self.elements.intersection(other.elements))

    def __sub__(self, other):
        return CustomSet(self.elements.difference(other.elements))

    def __add__(self, other):
        return CustomSet(self.elements.union(other.elements))

File: solutions/ruby/custom-set/1/custom_set.rb
class CustomSet
  def initialize(elements)
    @elements = elements.uniq
  end

  def empty?
    @elements.empty?
  end

  def member?(element)
    @elements.include?(element)
  end

  def subset?(other)
    @elements.all? { |e| other.member?(e) }
  end

  def disjoint?(other)
    @elements.none? { |e| other.member?(e) }
  end

  def eql?(other)
    @elements.sort == other.to_a.sort
  end

  def ==(other)
    eql?(other)
  end

  def add(element)
    @elements << element unless @elements.include?(element)
    self
  end

  def intersection(other)
    CustomSet.new(@elements.select { |e| other.member?(e) })
  end

  def difference(other)
    CustomSet.new(@elements.reject { |e| other.member?(e) })
  end

  def union(other)
    CustomSet.new(@elements + other.to_a)
  end

  def to_a
    @elements.dup
  end
end
